shell/browser/api/electron_api_web_contents.cc

```diff
@@ -1387,7 +1387,13 @@ void WebContents::HandleNewRenderFrame(
     rwh_impl->disable_hidden_ = !background_throttling_;
 
   auto* web_frame = WebFrameMain::FromRenderFrameHost(render_frame_host);
-  if (web_frame)
+  // When render process reuse is disabled a new siteinstance will always be
+  // forced for every navigation, if a WebFrameMain instance was created
+  // for a FrameTreeNodeId before navigation started, the corresponding
+  // RenderFrameHost will not be the same when the navigation completes.
+  // Compare GlobalFrameRoutingId to avoid incorrect behavior.
+  if (web_frame && web_frame->render_frame_host()->GetGlobalFrameRoutingId() ==
+                       render_frame_host->GetGlobalFrameRoutingId())
```

Comment:
I think we may want to check here whether the incoming RFH is a speculative frame.

From the stack trace:
```
content::RenderFrameHostManager::CreateSpeculativeRenderFrame [0x00007FF7249C0364+512] (o:\content\browser\renderer_host\render_frame_host_manager.cc:2651)
content::RenderFrameHostManager::CreateSpeculativeRenderFrameHost [0x00007FF7249BCFFC+244] (o:\content\browser\renderer_host\render_frame_host_manager.cc:2596)
```

We do some filtering for this in the latest code: https://github.com/electron/electron/blob/f39ba9281c524bd6d8f70b0724d6bcedbb499aa6/shell/browser/api/electron_api_web_contents.cc#L1397-L1407

And if we still want `Connect()` to be called for this WFM instance, we may need to listen for when the RFH transitions to its Active lifecycle state in `WebContentsObserver::RenderFrameHostStateChanged()`.

As mentioned in the referenced code, eventually RenderFrameCreated won't be called for speculative frames (https://crbug.com/1183639).

Reply:
We always abandon speculative hosts for new navigation in old process architecture https://github.com/electron/electron/blob/e74eac61ae317e095a301508fdb1c31ebc8a0f0b/shell/browser/electron_browser_client.cc#L647-L695, it should be fine to not connect in that case.